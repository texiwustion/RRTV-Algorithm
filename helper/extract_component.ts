import * as j from 'jscodeshift'

/**
 * structure: VD (RS (JF (JE (JE JE))))
 */
const text = `const App = () => {
    return (
        <>
            <div>
                <button id="3"></button>
                <p>this is a para</p>
            </div>
        </>
    )
}`

/**
 * the range of selection, present by offset
 * HERE: exactly a whole <div> in text
 */
const range = { start: 56, end: 159 }

export default function transformer(file: j.FileInfo, api: j.API) {
    const j: j.JSCodeshift = api.jscodeshift;
    const root: j.Collection = j(file.source);

    const checker = new Checker()

    {
        const jsxElements: j.Collection = root.findJSXElements()
        console.log('line:22 jsxElements::: ', jsxElements);

        const bestJSXElement = filterMethod.range.bestJSXElement(jsxElements)
        console.log('line:34 bestJSXElement::: ', bestJSXElement);
    }

    {
        const returnStatements: j.Collection = root.find(j.ReturnStatement)
        console.log('line:27 returnStatements::: ', returnStatements);
    }

    {
        const variableDeclaration: j.Collection = root.find(j.VariableDeclaration)
        console.log('line:32 variableDeclaration::: ', variableDeclaration);
    }
    return root.toSource();
}

interface checkProps {
    check: boolean;
    cache?: any;
}

class Checker {
    get(path: j.ASTPath, attr: string): checkProps {
        const res: any = path.get(attr)
        if (!res) {
            return { check: false }
        }
        return { check: true, cache: res }
    }
    range(path: j.ASTPath, range: any): checkProps {
        // if (node?.start <= range.start && range.end <= node?.end) {
        //  node 上无 start 属性
        // }
        const pathStart = path.get('start')
        const pathEnd = path.get('end')
        if (!pathStart || !pathEnd) {
            return { check: false }
        }
        if (pathStart <= range.start && range.end <= pathEnd) {
            return { check: true }
        }
        return { check: false }
    }
}

const filterMethod = {
    range: {
        bestJSXElement: (jsxElements: j.Collection): j.ASTPath | boolean => {
            let best: j.ASTPath | undefined = undefined
            let range = globalThis.range
            jsxElements.forEach((path: j.ASTPath) => {
                if (new Checker().range(path, range)) {
                    best = path
                    range = {
                        start: path.get('start'),
                        end: path.get('end')
                    }
                }
            })
            if (!best)
                return false
            return best
        }
    }
}