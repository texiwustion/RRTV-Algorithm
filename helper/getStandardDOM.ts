export const getStandardDOM = {
    byText(text: string) {
        const openingTagRegex = /<[^/].*?>/g; // 匹配开口语句
        const closingTagRegex = /<\/.*?>/g; // 匹配闭合语句

        const openingMatches = text.match(openingTagRegex); // 所有开口语句的匹配结果
        const closingMatches = text.match(closingTagRegex); // 所有闭合语句的匹配结果

        if (openingMatches === null || closingMatches === null) {
            console.log("这段文本内无完整 DOM 元素！")
            return
        }

        const firstOpeningMatch = openingMatches[0]
        const lastClosingMatch = closingMatches[closingMatches.length - 1]
        const startIndex: number = text.indexOf(firstOpeningMatch) - 1
        const endIndex: number = text.lastIndexOf(lastClosingMatch) + lastClosingMatch.length

        const rawText = text.slice(startIndex, endIndex)
        console.log('line:20 rawText::: ', rawText);
        const newText = "<>\n" + rawText + "\n</>"
        console.log('line:22 newText::: ', newText);
    }
}

const text = `=> <button> eiarsnt</button> 
<div>
    <p></p>
    <button> </button>
</div`

getStandardDOM.byText(text)