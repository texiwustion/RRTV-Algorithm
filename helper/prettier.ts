import { Options } from "prettier";
import * as prettier from "prettier";

const options: Options = {
    trailingComma: 'es5',
    tabWidth: 4,
    semi: true,
    singleQuote: true,
    jsxSingleQuote: true,
    arrowParens: "always",
    parser: "babel",
    plugins: [require.resolve("prettier/parser-babel")]
}
export function prettierCode(code: string) {
    try {
        // 参数1：代码字符串，参数2：格式化配置
        const res = prettier.format(code, options);
        return res
    } catch (error) {
        // 如果格式化失败,返回源码
        return code;
    }
}
