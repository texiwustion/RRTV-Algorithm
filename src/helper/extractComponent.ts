import j, {
    ASTPath,
    Collection,
    blockStatement,
    functionDeclaration,
    identifier,
    jsxClosingFragment,
    jsxFragment,
    jsxOpeningFragment,
    returnStatement,
} from "jscodeshift";

export type Range = {
    start: number;
    end: number;
};

const PRINT = (output: any) => {
    console.log(output);
};

export type DomElement = string | null;
export type Props<T> = {
    result: T;
    data: Object;
};
export type extractDomElementProps = Props<DomElement>;
export type checkProps = Props<boolean>;
export default class extractComponent {
    private domElement: DomElement;
    private text: string;
    private checkResult: boolean;
    private deltaRange: Range;
    private range: Range;
    private source: string;
    extractDomElement(text: string): extractDomElementProps {
        const openingTagRegex = /<[^/].*?>/g; // 匹配开口语句
        const closingTagRegex = /<\/.*?>/g; // 匹配闭合语句

        const openingMatches = text.match(openingTagRegex); // 所有开口语句的匹配结果
        const closingMatches = text.match(closingTagRegex); // 所有闭合语句的匹配结果

        if (openingMatches === null || closingMatches === null) {
            console.log("这段文本内无完整 DOM 元素！");
            return {
                result: null,
                data: {},
            };
        }

        const firstOpeningMatch = openingMatches[0];
        const lastClosingMatch = closingMatches[closingMatches.length - 1];
        const startIndex: number = text.indexOf(firstOpeningMatch) - 1;
        const endIndex: number =
            text.lastIndexOf(lastClosingMatch) + lastClosingMatch.length - 1;

        const domElement: DomElement = text.slice(startIndex, endIndex + 1);

        return {
            result: domElement,
            data: {
                deltaRange: {
                    start: startIndex,
                    end: endIndex,
                },
            },
        };
    }
    withFragement(domElement: DomElement): DomElement {
        return "<>\n" + domElement! + "\n</>";
    }

    check(): checkProps {
        if (this.domElement === null) {
            return {
                result: false,
                data: {},
            };
        }

        let root: Collection | null = null;
        let validDomElement: DomElement = null;
        try {
            validDomElement = this.domElement;
            root = j(validDomElement);
        } catch (e) {
            console.log("[extractComponent] 所选区域内可能存在兄弟DOM元素");
            try {
                validDomElement = this.withFragement(this.domElement);
                root = j(validDomElement);
            } catch (e) {
                console.log("[extractComponent] 所选区域内无有效DOM元素");
                return {
                    result: false,
                    data: {},
                };
            }
        }

        return {
            result: true,
            data: {
                validDomElement,
            },
        };
    }
    handleSource(source: string, range: Range) {
        const text: string = source.slice(range.start, range.end);
        return text;
    }
    run(source: string, range: Range) {
        this.text = this.handleSource(source, range);

        {
            const { result, data } = this.extractDomElement(this.text);
            this.domElement = result;
            this.deltaRange = data["deltaRange"] ?? null;
        }

        {
            const { result, data } = this.check();
            this.checkResult = result;
        }

        if (this.deltaRange !== null)
            this.range = {
                start: range.start + this.deltaRange.start,
                end: range.start + this.deltaRange.end,
            };

        this.source = this.withReplacedJSXElement(source);
        return this.checkResult;
    }
    modify(): string {
        const fragment = jsxFragment(
            jsxOpeningFragment(),
            jsxClosingFragment(),
        );
        const findJSXElement = j(this.domElement).find(j.JSXElement);
        const newReturnBody = (
            findJSXElement.size
                ? findJSXElement
                : j(this.domElement).find(j.JSXFragment)
        ).get(0).node;
        console.log(newReturnBody);
        const newFunction = functionDeclaration(
            identifier("NewFunction"),
            [],
            blockStatement([returnStatement(newReturnBody)]),
        );

        const root: Collection = j(this.source);
        const insertPath: ASTPath = root
            .findJSXElements("NewFunction")
            .closestScope()
            .paths()[0].parent.parent;
        insertPath.insertBefore(newFunction);
        return root.toSource();
    }
    withReplacedJSXElement(source: string) {
        return new Array(
            source.substring(0, this.range.start - 1),
            "\t<NewFunction />",
            source.substring(this.range.end + 1),
        ).join("");
    }
}
