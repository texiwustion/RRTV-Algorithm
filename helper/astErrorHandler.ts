import { getStandardDOM } from "./getStandardDOM"
import * as j from 'jscodeshift'

/**
 * use babel analyser for making exactly complete jsxElement
 * @param text String, a range of text from USER selection
 * @returns null | root: j.Collection (a complete jsxElement)
 */
export function astHandle(text: string): any | null {
    const result = getStandardDOM.byText(text)
    console.log('line:11 result::: ', result);
    let useText: string = "rawText"
    if (result === null) {
        return null
    }
    let root: j.Collection<any> | null = null
    try {
        root = j(result.rawText)
    }
    catch (e) {
        try {
            root = j(result.newText)
            useText = "newText"
        }
        catch (e) {
            console.log("[astErrorHandler]Line24::: Error: 所选区域内 DOM 元素不完整！")
            return null
            /* <hello><button id="1" me="2"></button>*/
        }
    }

    if (root === null) {
        return null
    }
    else {
        return { root, ...result, useText }
    }
}
