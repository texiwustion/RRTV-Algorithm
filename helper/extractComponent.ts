import { astHandle } from './astErrorHandler';
import * as j from 'jscodeshift'

export const extractComponent = {
    jsxElement: null,
    rawText: null,
    newText: null,
    deltaRange: null,
    /**
     * check this text if is valid
     * @param text String, from USER selection
     * @returns boolean to judge this check if valid
     */
    check(text: string): boolean {
        const result = astHandle(text)

        if (result === null) {
            return false
        }
        else {
            this.jsxElement = result.root
            this.rawText = result.rawText
            this.newText = result.newText
            this.deltaRange = result.deltaRange
            return true
        }
    },
    modify(document: string) {
        if (this.jsxElement === null) {
            throw new Error("[extractComponent] Line 22 ::: 在 modify 前未执行 check!")
        }

        const root = j(document)
    }
}

export interface Range {
    start: number;
    end: number;
}

export function getNewRange(range: Range, deltaRange: Range) {
    return {
        start: range.start + deltaRange.start,
        end: range.start + deltaRange.end
    }
}