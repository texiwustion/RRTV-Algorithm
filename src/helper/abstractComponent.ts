import { extractComponent } from "./extractComponent";

export class abstractComponent {
    check = extractComponent.check;
}
