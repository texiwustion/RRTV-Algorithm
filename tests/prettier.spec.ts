import { prettierCode } from "../src/helper/prettier";

it("test prettier", async () => {
    const code = `//格式故意搞乱，不对齐！
function uglyCode () {
                    console.log("this is an ugly function!") 
 }`;
    expect(await prettierCode(code)).toMatchSnapshot();
});
