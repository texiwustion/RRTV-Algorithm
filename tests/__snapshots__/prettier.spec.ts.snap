// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test prettier 1`] = `
"//格式故意搞乱，不对齐！
function uglyCode() {
    console.log('this is an ugly function!');
}
"
`;
